File: Practice/AOJ/DSL/DSL1_A/dsl1_a.py
import sys
sys.setrecursionlimit(10**8)

n, q = list(map(int, input().split()))
arr = list(range(n))
rank = [0 for _ in range(n)]


def root(t):
    # 経路圧縮
    if t != arr[t]:
        arr[t] = root(arr[t])
    return arr[t]


def unite(x, y):
    rootx = root(x)
    rooty = root(y)
    # ランク付
    if rank[rootx] < rank[rooty]:
        arr[rooty] = rootx
    elif rank[rooty] < rank[rootx]:
        arr[rootx] = rooty
    elif rootx != rooty:
        arr[rooty] = rootx
        rank[rootx] += 1


def find(x, y):
    return root(x) == root(y)


for _ in range(q):
    command, x, y = list(map(int, input().split()))
    if command == 0:
        unite(x, y)
    else:
        print(1 if find(x, y) else 0)

File: Practice/AOJ/DSL/DSL1_A/dsl1_a.php
<?php
  $arr = array_map(intval, explode(" ", trim(fgets(STDIN))));
  $n = $arr[0];
  $q = $arr[1];
  $parents = range(0, $n-1);

  function root($x){
    global $parents;
    if($x === $parents[$x]){
      return $x;
    }else{
      return $parents[$x] = root($parents[$x]);
    }
  }

  function same($x, $y){
    return root($x) == root($y);
  }

  function union($x, $y){
    global $parents;
    if(!same($x, $y)){
      $parents[root($x)] = $y;
    }
  }

  for($i=0; $i<$q; $i++){
    $query = array_map(intVal, explode(" ", trim(fgets(STDIN))));
    if($query[0] === 0){
      // unite
      union($query[1], $query[2]);
    }else{
      // is same
      echo same($query[1], $query[2]) ? 1 . PHP_EOL : 0 . PHP_EOL;
    }
  }
?>